state/charm.go

```diff
@@ -506,48 +235,6 @@ func (st *State) findCharm(curl *charm.URL) (CharmRefFull, error) {
 	return fromInternalCharm(ch, curl.String()), nil
 }
 
-// LatestPlaceholderCharm returns the latest charm described by the
-// given URL but which is not yet deployed.
-func (st *State) LatestPlaceholderCharm(curl *charm.URL) (*Charm, error) {
-	charms, closer := st.db().GetCollection(charmsC)
-	defer closer()
-
-	noRevURL := curl.WithRevision(-1)
-	curlRegex := "^" + regexp.QuoteMeta(st.docID(noRevURL.String()))
-	var docs []charmDoc
-	err := charms.Find(bson.D{{"_id", bson.D{{"$regex", curlRegex}}}, {"placeholder", true}}).All(&docs)
-	if err != nil {
-		return nil, errors.Annotatef(err, "cannot get charm %q", curl)
-	}
-	// Find the highest revision.
-	var latest charmDoc
-	var latestURL *charm.URL
-	for _, doc := range docs {
-		if latest.URL == nil {
-			latest = doc
-			latestURL, err = charm.ParseURL(*latest.URL)
-			if err != nil {
-				return nil, errors.Annotatef(err, "latest charm url")
-			}
-		}
-		docURL, err := charm.ParseURL(*doc.URL)
-		if err != nil {
-			return nil, errors.Annotatef(err, "current charm url")
-		}
-		if docURL.Revision > latestURL.Revision {
```

Comment:
This is actually incorrect for charmhub charms, this should just be any row in the database that isn't associated with an application. We can't use revisions as a way to see if something is "newer".